File: C#/Solution #2  O(n) Time & O(n) Space.cs

// O(n) time | O(n) space

​using System.Collections.Generic;

public class Program {

	public static int GetNthFib(int n) {
		Dictionary<int, int> memoize = new Dictionary<int, int>();
		memoize.Add(1, 0);
		memoize.Add(2, 1);
		return GetNthFib(n, memoize);
	}

	public static int GetNthFib(int n, Dictionary<int, int> memoize) {
		if (memoize.ContainsKey(n)) {
			return memoize[n];
		} else {
			memoize.Add(n, GetNthFib(n - 1, memoize) + GetNthFib(n - 2, memoize));
			return memoize[n];
		}
	}
}

File: Java/Solution #2  O(n) Time & O(n) Space.java

// O(n) time | O(n) space

import java.util.HashMap;

class Program {

  public static int getNthFib(int n) {
    HashMap<Integer, Integer> memoize = new HashMap<Integer, Integer>();
    memoize.put(1, 0);
    memoize.put(2, 1);
    return getNthFib(n, memoize);
  }

  public static int getNthFib(int n, HashMap<Integer, Integer> memoize) {
    if (memoize.containsKey(n)) {
      return memoize.get(n);
    } else {
      memoize.put(n, getNthFib(n - 1, memoize) + getNthFib(n - 2, memoize));
      return memoize.get(n);
    }
  }
}
